### PR TITLE
Validate unary ops in WGSL lowerer

### PR DESCRIPTION
**Pull request by BenjaminBrienen**
**Connections**
No connections.

**Description**
I noticed that naga accepts invalid WGSL while playing with wesl-rs and wgsl-analyzer. The original issue is that it accepted `-mat2x2f()` (negating a matrix) as a valid expression. However, according to the WGSL specification, that violates the precondition.

**Testing**
Added tests to assert the error messages.

**Squash or Rebase?**

Either.

**Checklist**

- [x] I self-reviewed and fully understand this PR.
- [x] WebGPU implementations built with `wgpu` may be affected behaviorally.
- [ ] Validation and feature gates are in place to confine behavioral changes.
- [x] Tests demonstrate the validation and altered logic works. 
- [ ] `CHANGELOG.md` entries for the user-facing effects of this change are present.
- [x] The PR is minimal, and doesn't make sense to land as multiple PRs.
- [x] Commits are logically scoped and individually reviewable.
- [x] The PR description has enough context to understand the motivation and solution implemented.

### CODE REVIEW

File: naga/src/front/wgsl/lower/mod.rs [hunk 1]
```diff
@@ -2909,6 +2906,76 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
...
+            op,
+            operand_type,
+        };
+
+        let expr = self.expression_for_abstract(expr, ctx)?;
+        ctx.grow_types(expr)?;
+
+        let expr_resolution = resolve!(ctx, expr);
+        // All unary operators are only defined for scalars and vectors of scalars.
+        let size_and_scalar = expr_resolution
```
Comment (ErichDonGubler) [inline naga/src/front/wgsl/lower/mod.rs]: nit: `size_and_scalar` isn't useful to keep here, so let's chain `map` on what it was based on instead.

Comment (ErichDonGubler) [reply naga/src/front/wgsl/lower/mod.rs]: Implemented as suggested with 5872c2ca.

File: naga/src/front/wgsl/lower/mod.rs [hunk 2]
```diff
@@ -2909,6 +2906,76 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
...
+        let make_error = |operand_type: String| Error::InvalidUnaryOperandType {
+            span,
+            op,
+            operand_type,
+        };
+
+        let expr = self.expression_for_abstract(expr, ctx)?;
+        ctx.grow_types(expr)?;
+
+        let expr_resolution = resolve!(ctx, expr);
```
Comment (ErichDonGubler) [inline naga/src/front/wgsl/lower/mod.rs]: suggestion(non-blocking): s/expr_resolution/expr_ty_resolution/, just in case somebody was unfamiliar with `resolve!(…)` like I was.

Comment (ErichDonGubler) [reply naga/src/front/wgsl/lower/mod.rs]: Implemented as suggested with 55d5557c.

**Comment by BenjaminBrienen** [comment]
Note that the validator already catches some cases, but not all.
```
let x = mat2x2f();
let y = -x;
```
vs
```
let x = -mat2x2f();
```

File: naga/src/front/wgsl/lower/mod.rs [hunk 3]
```diff
@@ -2909,6 +2906,76 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
...
+        let size_and_scalar = expr_resolution
+            .inner_with(&ctx.module.types)
+            .vector_size_and_scalar();
+
+        let Some(kind) = size_and_scalar.map(|(_, scalar)| scalar.kind) else {
+            let operand_type = ctx.type_resolution_to_string(expr_resolution);
+            return Err(Box::new(make_error(operand_type)));
+        };
+        // validate preconditions
+        match op {
```
Comment (ErichDonGubler) [inline naga/src/front/wgsl/lower/mod.rs]: nit: This `match` seems to have a lot of duplicated error handling.

suggestion: If we match on `(op, kind)` instead, we can use patterns in a single place, and just relegate error handling to a single rest arm (i.e., `_ => { … }`), rather than duplicating for every match arm.

Comment (ErichDonGubler) [reply naga/src/front/wgsl/lower/mod.rs]: Implemented as suggested with 4cff8c22.

File: naga/src/front/wgsl/error.rs [hunk 4]
```diff
@@ -1023,6 +1030,20 @@ impl<'a> Error<'a> {
                         ),
                         "is missing a template argument"
                     ),
+                    Error::InvalidUnaryOperandType { op, operand_type, .. } => {
+                        let operator = match op {
+                            UnaryOperator::Negate => "-",
+                            UnaryOperator::LogicalNot => "!",
+                            UnaryOperator::BitwiseNot => "~",
+                        };
```
Comment (ErichDonGubler) [inline naga/src/front/wgsl/error.rs:1038]: thought: It'd be nice if we could better share this relation between unary operators and their string sigils. That seems out-of-scope for this PR, though.

**Review by ErichDonGubler — CHANGES_REQUESTED** [review]
Hey there! Welcome to your first contribution to wgpu! ❤️ Glad to have you here. This broadly LGTM, though there's at least one thing that needs attention before we're done.

> …, according to the WGSL specification, [this issue] violates the precondition.

issue: If this violate(d) the WGSL spec., then there should be corresponding WebGPU CTS coverage that we enable to prove it. I don't consider this PR complete until we've addressed that.

suggestion: Let's either enable something in `cts_runner/test.lst`, or have an upstream CTS issue number proving we've done our due diligence for coverage here.


---

<sub>I used [Conventional Comments](https://conventionalcomments.org/) in this review! I hope they help with clarity and tone. 🙂</sub>

**Comment by ErichDonGubler** [comment]
I've pushed several `fixup!` commits, which I hope facilitate with the easier feedback I've left. If you have no objection to them, feel free to `Resolve` the correspondent review conversations.

**Comment by BenjaminBrienen** [comment]
The cts unary validation is already enabled but it wasn't caught because it uses the form that Naga's validator already catches. Do we want to add test cases to the cts that are semantically the same test, but different as an implementation detail?

**Comment by BenjaminBrienen** [comment]
No objections to the fixups

**Comment by ErichDonGubler** [comment]
> The cts unary validation is already enabled but it wasn't caught because it uses the form that Naga's validator already catches. Do we want to add test cases to the cts that are semantically the same test, but different as an implementation detail?

Yes! And, for the purpose of this PR, just having a track number in CTS upstream noting this present and linked here is sufficient, in that case.

**Comment by andyleiserson** [comment]
The gap before was in const evaluation. `let` variables are not const-evaluatable even when they have a const initializer, so `let y = -x` does not get evaluated, but the RHS of `let x = -mat2x2f();` does.

I think it would also be worth changing const evaluation to not accept matrices in this context, although it's unreachable from the existing frontends after the other changes in this PR.

https://github.com/gfx-rs/wgpu/blob/390f05b69acb1efc3855ab0b380838fe9f09260b/naga/src/proc/constant_evaluator.rs#L2727-L2730

> Do we want to add test cases to the cts that are semantically the same test, but different as an implementation detail?

I think it's worth submitting cases like this as feedback, but I'd disagree that the CTS is uniformly expected to find bugs that are corner cases arising from implementation details. Although in this case, I would frame it as, the CTS tested this for runtime expressions but not const expressions, and having CTS tests for all the evaluation rules in const context does seem appropriate.

**Comment by BenjaminBrienen** [comment]
https://github.com/gpuweb/cts/issues/4694

**Comment by ErichDonGubler** [comment]
@andyleiserson:

> The gap before was in const evaluation. `let` variables are not const-evaluatable even when they have a const initializer, so `let y = -x` does not get evaluated, but the RHS of `let x = -mat2x2f();` does.
> 
> I think it would also be worth changing const evaluation to not accept matrices in this context, although it's unreachable from the existing frontends after the other changes in this PR.

Just pushed [f6da13d](https://github.com/gfx-rs/wgpu/pull/10036/commits/f6da13d586dceb266dcd3417c61f07056e486828). WDYT?

**Review by ErichDonGubler — APPROVED** [review]
Approving. I want to get consensus with @andyleiserson on the addition he suggested in the constant evaluator, but then I think this is good to merge after.

**Comment by BenjaminBrienen** [comment]
I didn't realize that changing `naga/src/proc/constant_evaluator.rs` was on the table. I'm surprised that other frontends don't use it.

**Comment by ErichDonGubler** [comment]
@BenjaminBrienen: It makes sense from a historical perspective; the constant evaluator was made to satisfy required semantics for WGSL, and it likely wouldn't do the correct thing for other languages' semantics as it currently is.

Incidentally, some of the `ConstantEvaluator` API is used by the `glsl` front-end, but not in the same way as WGSL.

**Review by andyleiserson — APPROVED** [review]
The change to the constant evaluator is what I was suggesting, thanks.

The constant evaluator is used by the GLSL frontend, and GLSL does allow negation of a matrix, but the transformation to `-1 * matrix` (because Naga IR, like WGSL, does not support it) is done before constant evaluation. I would generally prioritize aligning the constant evaluator with WGSL and/or Naga IR over concerns of GLSL or hypothetical other frontends.
